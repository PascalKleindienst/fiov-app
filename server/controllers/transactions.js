// ========================================
// TRANSACTIONS CONTROLLER=================
// ========================================

// List all transactions of the current user.
module.exports.list_all_transactions = (req, res) => {
    req.app.locals.models.Transactions.findAll({ //findAndCountAll
        where: { user_id: req.user.id },
        include: ['category']
    })
        .then( (transactions) => {
            res.json(transactions);
        })
        .catch((err) => {
            res.send(err);
        });
};

// Create a new transaction for the current user.
module.exports.create_transaction = (req, res) => {
    const { description = '', note = '', amount = 0, createdAt = null } = req.body;

    req.app.locals.models.Transactions
        .create({ description, note, amount, createdAt, user_id: req.user.id })
        .then((transaction) => {
            res.json(transaction.get());
        })
        .catch((err) => {
            res.send(err);
        });
};
