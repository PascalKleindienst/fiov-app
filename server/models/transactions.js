// ========================================
// Transactions Model =====================
// ========================================
module.exports = (sequelize, DataTypes) => {
    const Transactions = sequelize.define('Transactions', {
        description: DataTypes.STRING,
        note: DataTypes.TEXT,
        amount: DataTypes.INTEGER
    }, {});

    Transactions.associate = function(models) {
        Transactions.belongsTo(models.Categories, {
            foreignKey: 'category_id',
            onDelete: 'CASCADE',
            as: 'category'
        });

        Transactions.belongsTo(models.User, {
            foreignKey: 'user_id',
            onDelete: 'CASCADE',
            as: 'user'
        });
    };
    
    return Transactions;
};