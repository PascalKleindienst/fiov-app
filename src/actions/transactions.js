// ==============================================
// TRANSACTIONS ACTIONS =========================
// ==============================================
import ajax from 'fetchival';

// =============== Add Transaction ===============

// =============== Remove Transaction ===============

// =============== Update Transaction ===============

// =============== Set Transactions ===============
export const SET_TRANSACTIONS = 'SET_TRANSACTIONS';

export const setTransactions = () => {
    return (dispatch) => {
        return ajax('/api/transactions', { mode: 'cors', credentials: 'same-origin' }).get().then((transactions) => {
            dispatch({
                type: SET_TRANSACTIONS,
                transactions
            });
        });
    };
};