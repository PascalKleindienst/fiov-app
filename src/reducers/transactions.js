// ==============================================
// TRANSACTIONS REDUCERS ========================
// ==============================================
import { SET_TRANSACTIONS } from '../actions/transactions';

const transactionsReducerDefaultState = [];

export default (state = transactionsReducerDefaultState, action) => {
    switch (action.type) {
    case SET_TRANSACTIONS:
        return action.transactions;
    default:
        return state;
    }
};
