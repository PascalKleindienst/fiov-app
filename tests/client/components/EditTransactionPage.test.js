// ==============================================
// EDIT TRANSACTION PAGE TEST ===================
// ==============================================
// Mocks
const helpers = require('../../helpers');
import ajax from 'fetchival';
jest.mock('fetchival');

// Dependencies
import React from 'react';
import { shallow } from 'enzyme';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import EditTransactionPageContainer, { EditTransactionPage } from '../../../src/components/EditTransactionPage';
import transactions from '../fixtures/transactions';

const mockStore = configureMockStore([thunk]);

describe('my EditTransactionPage component', () => {
    let editTransaction, removeTransaction, translate, history, wrapper;

    beforeEach(() => {
        translate = jest.fn((msg) => msg);
        editTransaction = jest.fn();
        removeTransaction = jest.fn();
        history = { push: jest.fn() };
        wrapper = shallow(<EditTransactionPage editTransaction={editTransaction} removeTransaction={removeTransaction} translate={translate} history={history} transaction={transactions[1]} />);
    });

    test('should render EditTransactionPage correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should handle editTransaction', () => {
        wrapper.find('Connect(TransactionForm)').prop('onSubmit')(transactions[2]);

        expect(history.push).toHaveBeenLastCalledWith('/');
        expect(editTransaction).toHaveBeenLastCalledWith(transactions[1].id, transactions[2]);
    });
    
    test('should handle openModal', () => {
        wrapper.instance().handleOpenModal();
        expect(wrapper.instance().state).toEqual({ showModal: true });
    });

    test('should handle closeModal', () => {
        wrapper.instance().handleCloseModal();
        expect(wrapper.instance().state).toEqual({ showModal: false });
    });

    test('should handle removeTransaction', () => {
        wrapper.instance().onRemove();
        expect(history.push).toHaveBeenLastCalledWith('/');
        expect(removeTransaction).toHaveBeenLastCalledWith({ id: transactions[1].id });
    });

    test('should edit transaction on form submit', () => {
        const id = transactions[1].id;
        const store = mockStore({
            locale: { languages: ['en'] },
            transactions
        });
        const wrapper = shallow(<EditTransactionPageContainer store={store} history={history} match={ { params: { id } } } />);
        ajax.mockImplementation(() => ({ patch: helpers.resolvePromise({ id }) }));
        
        return wrapper.props().editTransaction(id, transactions[1]).then(() => {
            const actions = store.getActions();
            expect(actions[0]).toEqual({ type: 'EDIT_TRANSACTION', id,  updates: transactions[1] });
        });
    });
    
    test('should remove transaction on confirmation', () => {
        const id = transactions[1].id;
        const store = mockStore({
            locale: { languages: ['en'] },
            transactions
        });
        const wrapper = shallow(<EditTransactionPageContainer store={store} history={history} match={ { params: { id } } } />);
        ajax.mockImplementation(() => ({ delete: helpers.resolvePromise({ id }) }));
        
        return wrapper.props().removeTransaction({ id }).then(() => {
            const actions = store.getActions();
            expect(actions[0]).toEqual({ type: 'REMOVE_TRANSACTION', id });
        });
    });
});