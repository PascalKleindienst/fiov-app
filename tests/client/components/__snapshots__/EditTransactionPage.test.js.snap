// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`my EditTransactionPage component should render EditTransactionPage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h2
        className="page-header__title"
      >
        transactions.edit_transaction
      </h2>
    </div>
  </div>
  <div
    className="content-container"
  >
    <Connect(TransactionForm)
      onSubmit={[Function]}
      transaction={
        Object {
          "amount": -109500,
          "createdAt": "1969 12 28",
          "description": "Rent",
          "id": 2,
          "note": "",
        }
      }
    />
    <button
      className="button button--secondary button--remove"
      onClick={[Function]}
    >
      transactions.remove_transaction
    </button>
  </div>
  <Connect(DeleteTransactionModal)
    handleClose={[Function]}
    handleConfirmation={[Function]}
    isOpen={false}
  />
</div>
`;
